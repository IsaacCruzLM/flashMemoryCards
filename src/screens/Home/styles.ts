import {StyleSheet} from 'react-native';
// import themes from '../../styles/themes';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-between',
  },
});

export default styles;
