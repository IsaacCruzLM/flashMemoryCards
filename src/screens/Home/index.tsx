/* eslint-disable react-hooks/exhaustive-deps */
import React, {useEffect, useState, useContext} from 'react';
import {FlatList} from 'react-native';
import {withDatabase} from '@nozbe/watermelondb/DatabaseProvider';
import withObservables from '@nozbe/with-observables';
import {compose} from 'recompose';
import cloneDeep from 'lodash/cloneDeep';

import DefaultContainerView from '../../components/DefaultContainerView';
import NoteListCard from '../../components/NoteListCard';
import FloatingAddButton from '../../components/FloatingAddButton';
import EmpytListMessage from './components/EmpytListMessage';
import Dialog from '../../components/Dialog';
import {DEFAULT_FILTER_STATE, NotesFilter} from '../Note/List';

import noteNeedToBeRevised from '../../utils/noteValidations';
import useGetFromGlobalState from '../../hooks/useGetFromGlobalState';
import {NoteModelType} from '../../databases/models/noteModel';
import AppContext from '../../context/appContext';
import filterActions from '../../utils/filters';

import styles from './styles';
import {CardData} from '../Note/List/types';
import {rangeDataType} from '../../components/DataRangeInput/types';
import {filterState} from '../Note/List/types';

const Home = ({notes, categories, subjects, noteSubjects}: any) => {
  const {setFilterDialogOpenFunction} = useContext(AppContext);
  const [notesToRevise, setNotesToRevise] = useState([] as CardData[]);
  const [filters, setFilters] = useState(DEFAULT_FILTER_STATE as filterState);
  const searchQuery = useGetFromGlobalState('searchParams.Notes', '');
  const openFilterDialog = useGetFromGlobalState(
    'filterDialogOpen.Home',
    false,
  ) as boolean;

  const filterData = (note: NoteModelType) => {
    const {
      creationDate,
      lastRevision,
      category: categoryFilter,
      subjects: subjectsFilter,
    } = cloneDeep(filters);
    let dontFilter = true;

    if (categoryFilter && dontFilter) {
      dontFilter = note.category.id === categoryFilter;
    }

    if (subjectsFilter.length > 0 && dontFilter) {
      const subjectsIds = noteSubjects
        .filter((record: any) => record.note.id === note.id)
        .map((record: any) => record.subject.id);
      const someIdIsIncluded = subjectsIds.some((id: string) =>
        subjectsFilter.includes(id),
      );
      dontFilter = someIdIsIncluded;
    }

    if ((creationDate.init || creationDate.end) && dontFilter) {
      dontFilter = filterActions.rangeInputDateVerify(
        creationDate,
        note.createdAt.toString(),
      );
    }

    if ((lastRevision.init || lastRevision.end) && dontFilter) {
      dontFilter = filterActions.rangeInputDateVerify(
        lastRevision,
        note.lastRevision.toString(),
      );
    }

    return dontFilter;
  };

  useEffect(() => {
    const fetchNotes = async () => {
      console.log(notes);
      if (notes.length > 0) {
        const notesDataToRevise = [] as Array<CardData>;
        await Promise.all(
          notes
            .filter(noteNeedToBeRevised)
            .filter(filterData) // Filtro manual aqui
            .map(async (note: NoteModelType) => {
              const categoryName = (await note.category.fetch()).name;
              const subjectsFetch = await note.subjects.fetch();
              if (note.name.includes(searchQuery)) {
                notesDataToRevise.push({
                  id: note.id,
                  title: note.name,
                  creationDate: new Date(note.createdAt).toLocaleDateString(
                    'pt-BR',
                  ),
                  lastRevisionDate: new Date(
                    note.lastRevision,
                  ).toLocaleDateString('pt-BR'),
                  noteType: 'Texto',
                  category: categoryName,
                  subjects: (subjectsFetch || []).map(({name, color}: any) => ({
                    content: name,
                    color,
                  })),
                  nextRevisionDate: '',
                });
              }
            }),
        );

        setNotesToRevise(notesDataToRevise);
      }
    };

    fetchNotes();
  }, [categories, notes, noteSubjects, searchQuery, filters]);

  const handleFilterChange = (
    key: 'category' | 'subjects' | 'creationDate' | 'lastRevision',
    value: string & string[] & rangeDataType,
  ) => {
    const newFiltersClone = cloneDeep(filters);
    newFiltersClone[key] = value;
    setFilters(newFiltersClone);
  };

  const hideFilterDialog = () => setFilterDialogOpenFunction({Home: false});

  const resetFilterDialog = () => {
    setFilters(DEFAULT_FILTER_STATE as filterState);
    hideFilterDialog();
  };

  return (
    <DefaultContainerView>
      <FlatList
        testID="home-list"
        data={notesToRevise}
        contentContainerStyle={styles.listContainer}
        keyExtractor={item => item.id}
        renderItem={({item}) => {
          const {
            title,
            creationDate,
            lastRevisionDate,
            noteType,
            category,
            subjects: subjectsData,
          } = item;

          return (
            <NoteListCard
              title={title}
              creationDate={creationDate}
              lastRevisionDate={lastRevisionDate}
              noteType={noteType}
              category={category}
              subjects={subjectsData}
            />
          );
        }}
        style={styles.flatList}
        ListEmptyComponent={() => <EmpytListMessage />}
      />
      <FloatingAddButton routeName="NewNote" />
      <Dialog
        actions={[
          {
            label: 'Limpar',
            buttonMode: 'outlined',
            buttonAction: () => resetFilterDialog(),
          },
          {
            label: 'Filtrar',
            buttonMode: 'contained',
            buttonAction: () => hideFilterDialog(),
          },
        ]}
        isVisible={openFilterDialog}
        hideDialog={() => hideFilterDialog()}
        title={'Adicionar filtros'}>
        <NotesFilter
          filters={filters}
          handleFilterChange={handleFilterChange}
          categories={categories}
          subjects={subjects}
        />
      </Dialog>
    </DefaultContainerView>
  );
};

export default compose(
  withDatabase,
  withObservables([], ({database}: any) => {
    return {
      notes: database.get('notes').query(),
      categories: database.get('categories').query(),
      subjects: database.get('subjects').query(),
      noteSubjects: database.get('note_subjects').query(),
    };
  }),
)(Home);
