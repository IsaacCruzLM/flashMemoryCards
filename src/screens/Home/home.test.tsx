import React from 'react';
import renderer from 'react-test-renderer';
import Home from './index';

// Mock navigation
const mockNavigate = jest.fn();
const mockNavigation = {
  navigate: mockNavigate,
  goBack: jest.fn(),
};

// Mock Date Picker
jest.mock('react-native-date-picker', () => {
  return jest.fn().mockImplementation(({onConfirm, onCancel}) => {
    return {
      render: () => null,
      onConfirm,
      onCancel,
    };
  });
});

// Mock any potential Redux actions/hooks if you're using them
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => mockNavigation,
}));

jest.mock('react-native-paper', () => {
  const ActualReactNativePaper = jest.requireActual('react-native-paper');
  return {
    ...ActualReactNativePaper,
    Modal: ({children}: {children: React.ReactNode}) => <>{children}</>, // Simplesmente renderize os filhos, sem animações ou funcionalidade real
  };
});

describe('Home Screen', () => {
  let tree: renderer.ReactTestRenderer;
  beforeEach(() => {
    // Reset mocks before each test
    jest.clearAllMocks();

    tree = renderer.create(<Home navigation={mockNavigation} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  it('should render correctly', async () => {
    expect(tree.toJSON()).toBeTruthy();
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
