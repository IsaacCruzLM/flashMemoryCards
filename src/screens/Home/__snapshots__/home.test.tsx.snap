// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "space-between",
        "paddingHorizontal": 20,
        "paddingVertical": 24,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={Object {}}
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
        "width": "100%",
      }
    }
    testID="home-list"
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            Object {
              "marginVertical": 32,
            }
          }
        >
          <Icon
            color="#22272E"
            name="file-document-edit-outline"
            size={80}
          />
        </View>
        <Text
          style={
            Object {
              "color": "#22272E",
              "fontFamily": "System",
              "fontSize": 28,
              "fontWeight": "700",
              "marginBottom": 8,
              "textAlign": "center",
            }
          }
        >
          Nenhuma anotação a ser revisada
        </Text>
        <Text
          style={
            Object {
              "color": "#000000B3",
              "fontFamily": "System",
              "fontSize": 20,
              "fontWeight": "600",
              "marginBottom": 8,
              "textAlign": "center",
            }
          }
        >
          Insira novas anotações ou revise as anotações acessando-as pela seção de categorias
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#22272E",
        "borderRadius": 30,
        "bottom": 24,
        "height": 60,
        "justifyContent": "center",
        "position": "absolute",
        "right": 20,
        "width": 60,
      }
    }
    testID="floating-add-container"
  >
    <Icon
      color="#FFFFFF"
      name="plus"
      size={50}
    />
  </View>
</View>
`;
